nexus/src/sagas.rs

```diff
@@ -308,3 +335,332 @@ async fn sic_instance_ensure(
         .map(|_| ())
         .map_err(ActionError::action_failed)
 }
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct ParamsDiskCreate {
+    pub project_id: Uuid,
+    pub create_params: params::DiskCreate,
+}
+
+#[derive(Debug)]
+pub struct SagaDiskCreate;
+impl SagaType for SagaDiskCreate {
+    type SagaParamsType = Arc<ParamsDiskCreate>;
+    type ExecContextType = Arc<SagaContext>;
+}
+
+fn saga_disk_create() -> SagaTemplate<SagaDiskCreate> {
+    let mut template_builder = SagaTemplateBuilder::new();
+
+    template_builder.append(
+        "disk_id",
+        "GenerateDiskId",
+        new_action_noop_undo(sdc_generate_uuid),
+    );
+
+    template_builder.append(
+        "created_disk",
+        "CreateDiskRecord",
+        ActionFunc::new_action(
+            sdc_create_disk_record,
+            sdc_create_disk_record_undo,
+        ),
+    );
+
+    template_builder.append(
+        "datasets_and_regions",
+        "AllocRegions",
+        ActionFunc::new_action(sdc_alloc_regions, sdc_alloc_regions_undo),
+    );
+
+    template_builder.append(
+        "regions_ensure",
+        "RegionsEnsure",
+        ActionFunc::new_action(sdc_regions_ensure, sdc_regions_ensure_undo),
+    );
+
+    template_builder.append(
+        "disk_runtime",
+        "FinalizeDiskRecord",
+        new_action_noop_undo(sdc_finalize_disk_record),
+    );
+
+    template_builder.build()
+}
+
+async fn sdc_generate_uuid(
+    _: ActionContext<SagaDiskCreate>,
+) -> Result<Uuid, ActionError> {
+    Ok(Uuid::new_v4())
+}
+
+async fn sdc_create_disk_record(
+    sagactx: ActionContext<SagaDiskCreate>,
+) -> Result<db::model::Disk, ActionError> {
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params();
+
+    let disk_id = sagactx.lookup::<Uuid>("disk_id")?;
+
+    // NOTE: This could be done in a transaction alongside region allocation?
+    //
+    // Unclear if it's a problem to let this disk exist without any backing
+    // regions for a brief period of time, or if that's under the valid
+    // jurisdiction of "Creating".
+    let disk = db::model::Disk::new(
+        disk_id,
+        params.project_id,
+        params.create_params.clone(),
+        db::model::DiskRuntimeState::new(),
+    );
+    let disk_created = osagactx
+        .datastore()
+        .project_create_disk(disk)
+        .await
+        .map_err(ActionError::action_failed)?;
+    Ok(disk_created)
+}
+
+async fn sdc_create_disk_record_undo(
+    sagactx: ActionContext<SagaDiskCreate>,
+) -> Result<(), anyhow::Error> {
+    let osagactx = sagactx.user_data();
+
+    let disk_id = sagactx.lookup::<Uuid>("disk_id")?;
+    osagactx.datastore().project_delete_disk_no_auth(&disk_id).await?;
```

Comment:
I think that for this to be idempotent, project_delete_disk_internal() should succeed if the disk's state is already "Destroyed" and time_deleted != NULL.  Does it?

That said, I think you want an attempt at the API to delete a Disk that's already been deleted to fail.  Maybe that already happens by virtue of the way it gets looked up?  But before this change there was an explicit check.